class.jetpack-post-images.php

```diff
@@ -328,21 +334,48 @@ static function from_html( $html_or_id ) {
 			return $images;
 		}
 
-		preg_match_all( '!<img.*src=[\'"]([^"]+)[\'"].*/?>!iUs', $html, $matches );
-		if ( !empty( $matches[1] ) ) {
-			foreach ( $matches[1] as $match ) {
-				if ( stristr( $match, '/smilies/' ) )
-					continue;
+		// Do not go any further if DOMDocument is disabled on the server.
+		if ( ! class_exists( 'DOMDocument' ) ) {
+			return $images;
+		}
 
-				$images[] = array(
-					'type'  => 'image',
-					'from'  => 'html',
-					'src'   => html_entity_decode( $match ),
-					'href'  => '', // No link to apply to these. Might potentially parse for that as well, but not for now
-				);
+		// Let's grab all image tags from the HTML.
+		$dom_doc = new DOMDocument;
+		$dom_doc->loadHTML( $html );
+		$image_tags = $dom_doc->getElementsByTagName( 'img' );
+
+		// For each image Tag, make sure it can be added to the $images array, and add it.
+		foreach ( $image_tags as $image_tag ) {
+			$img_src = $image_tag->getAttribute( 'src' );
+
+			if ( ! isset( $img_src ) || empty( $img_src ) ) {
```

Comment:
We can skip the isset check -- empty handles that internally, and it's always going to have something as we're explicitly setting it two lines above.  It may be an empty string, but it'll still be set.